lua/bufferline/highlight.lua

```diff
@@ -0,0 +1,95 @@
+-- !::exe [So]
+
+-------------------
+-- Section: helpers
+-------------------
+
+--- Generate a color.
+--- @param index string where to look for the color.
+--- @param groups table<string> the groups to source the color from.
+--- @param default string the foreground color to use if no `groups` have a valid color.
+--- @return string color
+local function color(index, groups, default)
+  for _, group in ipairs(groups) do
+    local hl = vim.api.nvim_get_hl_by_name(group, true)
+    if hl[index] then
+      return string.format('#%06x', hl[index])
+    end
+  end
+  return default
+end
+
+--- Generate a foreground color.
+--- @param groups table<string> the groups to source the foreground color from.
+--- @param default string the foreground color to use if no `groups` have a valid foreground color.
+--- @return string color
+local function fg(groups, default)
+  return color('foreground', groups, default)
+end
+
+--- Generate a background color.
+--- @param groups table<string> the groups to source the background color from.
+--- @param default string the background color to use if no `groups` have a valid background color.
+--- @return string color
+local function bg(groups, default)
+  return color('background', groups, default)
+end
+
+------------------
+-- Section: module
+------------------
+
+--- @class bufferline.highlight
+local highlight = {}
+
+-- Initialize highlights
+function highlight.setup()
+  local fg_target = 'red'
+
+  local fg_current  = fg({'Normal'}, '#efefef')
+  local fg_visible  = fg({'TabLineSel'}, '#efefef')
+  local fg_inactive = fg({'TabLineFill'}, '#888888')
+
+  local fg_modified = fg({'WarningMsg'}, '#E5AB0E')
+  local fg_special  = fg({'Special'}, '#599eff')
+  local fg_subtle = fg({'NonText', 'Comment'}, '#555555')
+
+  local bg_current  = bg({'Normal'}, 'none')
+  local bg_visible  = bg({'TabLineSel', 'Normal'}, 'none')
+  local bg_inactive = bg({'TabLineFill', 'StatusLine'}, 'none')
+
+  --      Current: current buffer
+  --      Visible: visible but not current buffer
+  --     Inactive: invisible but not current buffer
+  --        -Icon: filetype icon
+  --       -Index: buffer index
+  --         -Mod: when modified
+  --        -Sign: the separator between buffers
+  --      -Target: letter in buffer-picking mode
```

Comment:
Here maybe add `local set_hl = vim.api.nvim_set_hl`? Easier to read, less noise.